File: kotlin-desktop-toolkit-linux/src/main/kotlin/org/jetbrains/desktop/linux/Geometry.kt
package org.jetbrains.desktop.linux

import kotlin.math.roundToInt

public typealias PhysicalPixels = Int
public typealias LogicalPixels = Float

public data class PhysicalSize(
    val width: PhysicalPixels,
    val height: PhysicalPixels,
) {
    public companion object;
    public fun toLogical(scale: Float): LogicalSize = LogicalSize(width / scale, height / scale)
}

public data class PhysicalPoint(
    val x: PhysicalPixels,
    val y: PhysicalPixels,
) {
    public companion object;
    public fun toLogical(scale: Float): LogicalPoint = LogicalPoint(x / scale, y / scale)
}

public data class LogicalSize(
    val width: LogicalPixels,
    val height: LogicalPixels,
) {
    public companion object;
    public fun toPhysical(scale: Float): PhysicalSize = PhysicalSize((width * scale).roundToInt(), (height * scale).roundToInt())
}

public data class LogicalPoint(
    val x: LogicalPixels,
    val y: LogicalPixels,
) {
    public companion object {
        public val Zero: LogicalPoint = LogicalPoint(0f, 0f)
    }
    public fun toPhysical(scale: Float): PhysicalPoint = PhysicalPoint((x * scale).roundToInt(), (y * scale).roundToInt())
}

public data class LogicalRect(val point: LogicalPoint, val size: LogicalSize) {
    public fun contains(p: LogicalPoint): Boolean {
        return p.x > point.x &&
            p.x < point.x + size.width &&
            p.y > point.y &&
            p.y < point.y + size.height
    }
}

File: native/desktop-linux-x11/src/linux/geometry.rs
#[repr(transparent)]
#[derive(Debug, Clone, Copy)]
pub struct PhysicalPixels(pub i32);

#[repr(transparent)]
#[derive(Debug, Default, Clone, Copy)]
pub struct LogicalPixels(pub f64);

impl LogicalPixels {
    #[allow(clippy::cast_possible_truncation)]
    #[must_use]
    pub const fn round(&self) -> i32 {
        self.0.round() as i32
    }
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct PhysicalSize {
    pub width: PhysicalPixels,
    pub height: PhysicalPixels,
}

#[repr(C)]
#[derive(Debug, Default, Clone, Copy)]
pub struct LogicalPoint {
    pub x: LogicalPixels,
    pub y: LogicalPixels,
}

impl<T: Into<f64>> From<(T, T)> for LogicalPoint {
    fn from(value: (T, T)) -> Self {
        Self {
            x: LogicalPixels(value.0.into()),
            y: LogicalPixels(value.1.into()),
        }
    }
}

#[repr(C)]
#[derive(Debug, Default, Clone, Copy)]
pub struct LogicalSize {
    pub width: LogicalPixels,
    pub height: LogicalPixels,
}

#[must_use]
pub fn to_logical_value<T: Into<f64>>(value: T, scale: f64) -> LogicalPixels {
    LogicalPixels(value.into() / scale)
}

#[must_use]
fn to_physical_value(value: f64, scale: f64) -> PhysicalPixels {
    #[allow(clippy::cast_possible_truncation)]
    PhysicalPixels((value * scale).round() as i32)
}

#[must_use]
pub fn round_to_u32(value: f64) -> u32 {
    u32::try_from(value.round() as i64).unwrap()
}

impl LogicalSize {
    #[must_use]
    pub fn to_physical(&self, scale: f64) -> PhysicalSize {
        PhysicalSize {
            width: to_physical_value(self.width.0, scale),
            height: to_physical_value(self.height.0, scale),
        }
    }
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct PhysicalRect {
    pub x: PhysicalPixels,
    pub y: PhysicalPixels,
    pub width: PhysicalPixels,
    pub height: PhysicalPixels,
}

#[repr(C)]
#[derive(Debug, Clone, Copy)]
pub struct LogicalRect {
    pub x: LogicalPixels,
    pub y: LogicalPixels,
    pub width: LogicalPixels,
    pub height: LogicalPixels,
}

impl PhysicalRect {
    #[must_use]
    pub fn to_logical(&self, scale: f64) -> LogicalRect {
        LogicalRect {
            x: to_logical_value(self.x.0, scale),
            y: to_logical_value(self.y.0, scale),
            width: to_logical_value(self.width.0, scale),
            height: to_logical_value(self.height.0, scale),
        }
    }

    #[must_use]
    pub const fn as_size(&self) -> PhysicalSize {
        PhysicalSize {
            width: self.width,
            height: self.height,
        }
    }
}

impl LogicalRect {
    #[must_use]
    pub fn to_physical(&self, scale: f64) -> PhysicalRect {
        PhysicalRect {
            x: to_physical_value(self.x.0, scale),
            y: to_physical_value(self.y.0, scale),
            width: to_physical_value(self.width.0, scale),
            height: to_physical_value(self.height.0, scale),
        }
    }

    #[must_use]
    pub const fn as_size(&self) -> LogicalSize {
        LogicalSize {
            width: self.width,
            height: self.height,
        }
    }
}

impl PhysicalSize {
    #[must_use]
    pub fn to_logical(&self, scale: f64) -> LogicalSize {
        LogicalSize {
            width: to_logical_value(self.width.0, scale),
            height: to_logical_value(self.height.0, scale),
        }
    }
}
